query_sheet.rb

```diff
@@ -186,5 +208,126 @@ def add_errors_to_message(
   error_message += ad_message + "\n"
 end
 
+def candidates_by_attribute(candidates, attribute, one_to_many=false)
+  candidates.select do |candidate|
+    candidate[attribute] &&
+      candidate[attribute].length > 0
+  end.reduce({}) do |obj, candidate|
+    if obj[candidate[attribute]]
+      obj[candidate[attribute]] << candidate
+    else
+      obj[candidate[attribute]] = one_to_many ? [candidate] : candidate
+    end
+    obj
+  end
+end
+
+def find_diffs(imported_candidates)
+  imported_candidates_by_nationbuilder_id = candidates_by_attribute(
+    imported_candidates,
+    'nationbuilder_id',
+  )
+  sheets_data = assembly_district_sheets.reduce({}) do |obj, ad_sheet|
+    obj[ad_sheet['assembly_district']] = candidates_by_attribute(
+      read_sheet(ad_sheet['spreadsheet_id'], 'Candidate View'),
+      'Email',
+    )
+    obj
+  end
+  binding.pry
+end
+
+def column_to_letter(num)
+  char = (num % 26 + 65).chr
+  remainder = num/26
+  if num >= 26
+    column_to_letter(remainder - 1) + char
+  else
+    char
+  end
+end
+
+def get_ad_and_ed_from_cc_sunlight(address)
+  base_uri = 'https://ccsunlight.org/api/v1/address/'
+  uri = URI(base_uri + CGI.escape(address))
+  request = Net::HTTP::Get.new(path=uri)
+  response = Net::HTTP.start(uri.host, uri.port, use_ssl: true) do |http|
+    http.request(request)
+  end
+  ad, ed = JSON.parse(response.body).values_at('ad', 'ed')
+  {
+    ad: ad,
+    ed: ed,
+  }
+end
+
+def format_address(candidate)
+  unless candidate['primary_address1'] &&
+    candidate['primary_address1'].length > 0 &&
+    candidate['primary_city'] &&
+    candidate['primary_city'].length > 0 &&
+    candidate['primary_city'].match(/brooklyn|new york/i) &&
+    candidate['primary_state'] &&
+    candidate['primary_state'].length > 0 &&
+    candidate['primary_state'].match(/ny/i) &&
+    candidate['primary_zip'] &&
+    candidate['primary_zip'].length > 0 &&
+    candidate['primary_zip'].length == 5
+      return nil
```

Comment:
Perhaps add a TODO to move into separate method, and do something if invalid.

Reply:
not a bad idea